Write policies and endpoints for StripeCustomer

amyschools opened the pull request:
# What's in this PR?

Added Stripe Customer policy for `index(filter_by_id)` and `show` endpoints to ability.ex, stripe_customer_policy.ex, factory, and tests to stripe_customer_policy_test.

## References
Progress on: #453 
Progress on: #447

npendery commented on blueprint/api.apib:
```diff
@@ -91,30 +96,34 @@ Exchange credentials for `access_token`.
   + id: 1 (string, required)
 
-  + theme (enum[string], optional)
+### Create a user [POST]
```
Shouldn't this be in its own `User` section as `## User [/users/new]`?

thomascchen commented on blueprint/api.apib:
```diff
@@ -165,48 +181,61 @@ Exchange credentials for `access_token`.
 
 ## User Relationships Base (object)
 + include User Roles Relationships
 + include User Skills Relationships
```
Should we also have relationships for `categories`, `skills`, `organizations`, and `organization_memberships`?

JoshSmith commented:
@christopherstyles are you working on this one still?

christopherstyles commented:
@JoshSmith yyyep, still working on it.

JoshSmith commented:
@christopherstyles should we close this out due to #464?

christopherstyles commented:
@JoshSmith I think so, feel free to close this one up ;)

amyschools committed:
Write policy and tests for Stripe Customer
Add StripeCustomerPolicy.create?
Add StripeCustomerController.create and .show
Moved slugged routes to bottom of router, to allow for authenticated :show endpoints
Call stripe service to create customer (requires test API key in dev)
Add attributes to stripe_customer, including email and id which hide when not current user

joshsmith commented on test/policies/stripe_customer_policy_test.exs:
```diff
@@ -0,0 +1,36 @@
+
+    end
+
+    test "returns false when user id is not the StripeCustomer id" do
```
I think this is more when the user id is not the Stripe customer's `user_id`, right?

joshsmith commented on test/policies/stripe_customer_policy_test.exs:
```diff
@@ -0,0 +1,36 @@
+
+    test "returns true when user is an admin" do
+
+    end
```
Would put `admin` first like in the test above. Think about it as reducing privileges as you go.

begedin commented:
In accordance with our discussion, we don't want to be using index endpoints on any `Stripe____` controllers, so we should remove those from here. That means we should remove the can calls matching `:index?`, the `index?` methods from the policy module, as well as tests for those.

If there are any additions to the policy helper that are used only by the `index?` methods, those should be removed as well.

Once that's done and the tests pass, we can review and merge this.

For this specific one, while it's on a good track already, the #456 PR can serve as an example, since the policy is similar.

amyschools commented:
@begedin similar question to #451 - I'm getting the  `Expected truthy, got false` error for this test:
`1) test show? returns true when user is viewing their own information (CodeCorps.StripeCustomerPolicyTest)`
The formatting is exactly the same as on #456 so I'm not sure what's wrong with it.

begedin commented:
@amyschools In this one, the tests seem to be OK, but your policy is wrong.

``` elixir
def show?(%User{admin: true}, %StripeCustomer{}), do: true
def show?(%User{} = user, %User{} = current_user, %StripeCustomer{} = stripe_customer), do: stripe_customer.id == current_user.id
def show?(%User{}, %StripeCustomer{}), do: false
```

The second one wont ever be used. Nothing is calling it.

What you need instead of that line is something like:

``` Elixir
def show?(%User{id: current_user_id}, %StripeCustomer{user_id: user_id}), do: current_user_id == user_id
```

begedin reviewed (changes requested):
This looks good, but it only handles the `show?` part of the issue #453 

I say it's good to merge, but upon merge, you should update the description of #453 so it shows that the `create?` part still needs to be done.

EDIT: Already did the update of description of #453, just make sure you check the respective boxes in the description upon merge.

EDIT: Was going to approve, but accidentally hit "request changes" instead. Corrected.

begedin commented:
@amyschools This needs a minor fix I missed last time. Since I need the endpoints to continue work on stripe, I'll take over and repurpose this PR. No need to work on it further, but good job up to this point.

begedin commented:
@JoshSmith This is still a work in progress, but the actual creation of stripe customer is done. What is left to handle is:
- [x] Right now, the client does not need to send any token. We simpl create a StripeCustomer and store it with a user_id reference, so the stripe request should not "validation-fail". It might have to request a token and it might have to fail, though. Could use your thoughts on that.
- [x] If the stripe request fails for some other reason, this is unhandled
- [x] We are expected to have a stripe api key in config/dev. It can be a test key, but I'm not sure how we should add it. We might need something like "dot_env", or use the "secret file" approach phoenix apps usually use

Once all of this is done, we can merge and move on to #454 

All of these changes should go together with code-corps/code-corps-ember#689

joshsmith commented:
@begedin did not get around to this today, but a couple notes:

Environment variables are already handled in `develop` in `config.exs`, like so:

``` elixir
config :stripity_stripe,
  api_key: System.get_env("STRIPE_SECRET_KEY"),
  connect_client_id: System.get_env("STRIPE_PLATFORM_CLIENT_ID")
```

Tests should probably not mock `stripity-stripe`. We should consider either writing a Bypass-oriented `stripity-stripe-testing` library that can be used here, or we should simply take a similar approach to our Segment API code as outlined in José's post here: http://blog.plataformatec.com.br/2015/10/mocks-and-explicit-contracts/

joshsmith commented:
I doubt we would send a token for creating a `StripeCustomer`. Can't see the utility in that. I think we probably want to eagerly create the customer the first time someone goes to donate (lands on the donation page?), or maybe even just after signing up? However it's done, it should be idempotent. And we do need to handle failure cases (API being down?)

begedin commented:
Update:
- Creation process works now. Had to update the stripe adapter for StripeCustomer so it properly converts from a `%Stripe.Customer{}` struct into a `%Map{}`.
- Without us realising, our switch to `JaResource` has made it so `ChangesetView` is no longer used for validation errors. Instead, the internal `errors.json-api` renderer is used, provided by `JaSerializer`. This works, but the errors no longer have a status field, so our client-side error formatter is no longer working properly. I made some changes here and made the required changes client-side to get it to a useful state.
- `stripity-stripe` wasn't handling 400 errors correctly, so I made a PR to fix that. Until that's merged, this PR will fail, unless we use switch our `mix.exs` to use the commit in the PR. I did that locally, but didn't want to push to this PR yet. The change is:

``` Elixir
{:stripity_stripe, "~> 2.0.0-alpha.1", git: "https://github.com/code-corps/stripity-stripe", ref: "cd1515989953da46f817de72549be5e431f2ae7d"}
```

The last thing left to do here is 
- [x] Write tests for `stripe_customer_controller`. This should be mergeable at that point. We probably need to implement something to mock stripe, similar to segment.

begedin commented:
Additional note/todo:
- [x] stripity-stripe should be returning a datetime for the `created` field. It's returning an integer right now.

joshsmith commented:
@begedin this is now returning a `DateTime` with `2.0.0-alpha.2`.

joshsmith commented on lib/code_corps/stripe/in_memory.ex:
```diff
@@ -0,0 +1,24 @@
+        livemode: false,
+        metadata: %{}
+      }
+    end
```
I'm not sure this has gone far enough. In http://blog.plataformatec.com.br/2015/10/mocks-and-explicit-contracts/ it looks like the recommendation is to use behaviours and callbacks. That is, we seem to be lacking explicit contracts right now.

The boundary of testing the HTTP interactions should probably happen at the level of `stripity-stripe` and not here, so I don't think that will have to be added. But even in adjusting the small cases we have here, I ended up with the behavior getting out of sync because the contract is not explicit.

joshsmith replied:
Further, now that I look at the effort we're putting into this, I'm not convinced that this shouldn't be happening in `stripity-stripe` or a related testing library, lest we reproduce the whole surface area of that library in our mocked API.

begedin replied:
I agree that it's not great, but I couldn't think of a better way. We would not be recreating the whole of `stripity-stripe`, but we would basically be recreating the parts that we use.

I'm open to suggestions, but until the library itself offers some testing tools, I don't know what else to do. Maybe https://github.com/parroty/exvcr ?

joshsmith commented on lib/code_corps/stripe/adapters/stripe_customer.ex:
```diff
@@ -1,7 +1,11 @@
 
-  def params_from_stripe(%{} = stripe_map) do
-    stripe_map |> rename("id", "id_from_stripe")
+  def params_from_stripe(%Stripe.Customer{} = customer) do
```
I might call this `to_params` instead.

joshsmith commented on test/views/stripe_customer_view_test.exs:
```diff
@@ -0,0 +1,32 @@
+      "data" => %{
+        "attributes" => %{
+          "email" => stripe_customer.email,
+        },
```
This appears to be missing attributes (and relationships) that are in the API Blueprint but not here.

joshsmith commented on test/views/user_view_test.exs:
```diff
@@ -38,7 +40,10 @@ defmodule CodeCorps.UserViewTest do
+            "data" => %{"id" => slugged_route.id |> Integer.to_string, "type" => "slugged-route"}
+          },
+          "stripe-customer" => %{
+            "data" => %{"id" => stripe_customer.id |> Integer.to_string, "type" => "stripe-customer"}
```
Do we want to be rendering this relationship out regardless of authorization?

begedin replied:
It's just our own db id, not an actual stripe id, so it should not be important. I wouldn't want to confuse ember data with hiding potential relationships. Our actual stripe_customer view can hide the `id_from_stripe` based on authorization, if necessary.

joshsmith replied:
I can see the logic in that, but do we actually risk confusing `ember-data` by doing so? If we're manually refreshing the user on authentication (we are), then it shouldn't matter.

Even the existence of those relationships leaking to other people seems problematic to me. Will think on this and ask around some.

joshsmith commented on web/controllers/stripe_customer_controller.ex:
```diff
@@ -0,0 +1,35 @@
+  defp handle_stripe_response({:ok, stripe_response}, attributes, conn) do
+    stripe_response
+    |> CodeCorps.Stripe.Adapters.StripeCustomer.params_from_stripe
+    |> Map.merge(attributes)
```
How does this merging work? Are there attributes that are not sent to Stripe or transformed somehow that need merged?

begedin replied:
The client also sends the `user_id`

begedin replied:
Made this more explicit by moving it into an adapter function called `add_non_stripe_attributes`

joshsmith commented on web/policies/stripe_customer_policy.ex:
```diff
@@ -0,0 +1,12 @@
+  alias Ecto.Changeset
+
+  def create?(%User{id: current_user_id}, %Changeset{changes: %{user_id: user_id}}), do: current_user_id == user_id
+  def create?(%User{}, %Changeset{}), do: false
```
Been wondering if we should just use `(_, _)` to handle the `false` case.

begedin replied:
Considering how our `ability.ex` is written, it can never be anything other than a `%User{}` and a `%Changeset{}`.

joshsmith commented on web/router.ex:
```diff
@@ -84,10 +81,19 @@ defmodule CodeCorps.Router do
+    get "/:slug", SluggedRouteController, :show
+    get "/:slug/projects", ProjectController, :index
+    get "/:slug/:project_slug", ProjectController, :show
+  end
```
Why don't we reorder authenticated routes above this so we're not separating the same pipeline?

begedin replied:
Didn't consider it, but fairly obvious in hindsight. Done!

joshsmith commented on web/views/stripe_customer_view.ex:
```diff
@@ -0,0 +1,9 @@
+  use CodeCorps.Web, :view
+  use JaSerializer.PhoenixView
+
+  attributes [:email]
```
Mentioned above but missing attributes defined in API docs. `email` and `id_from_stripe` are hidden when the user is not the current user.

begedin replied:
I've added all the attributes listed in the API docs.

I dod not add the subscriptions, because there is at the moment no relationship between the stripe_customer and the stripe_subscription. There is a relationship between `user` and `stripe_subscription` in that a `stripe_subscription` belongs to a `user` and a `stripe_plan`. A `stripe_customer` also belongs to a `user`. 

Should I update the API docs, or should I update our relationships? I'm pretty sure we mentioned this and decided to go with the relationship being on the `user`, not the `stripe_customer`.

Also, regarding the attributes, should we really render all those attributes? What do we get from it, really? Unless I'm missing something, we don't actually use the data client-side. We could even potentially go as far as completely hiding the `stripe_customer` from the client and just adding some sort of `connected_to_stripe` flag on the user record. I'm really not sure what the best approach is here, really.

joshsmith commented on lib/code_corps/stripe/adapters/stripe_customer.ex:
```diff
@@ -1,11 +1,29 @@
     |> Map.take([:created, :currency, :delinquent, :id, :email])
     |> rename(:id, :id_from_stripe)
     |> keys_to_string
   end
```
What about making `to_params` take `customer` and `attributes` so we don't have to do two separate adapter operations, then this function will combine the `add_non_stripe_attributes` in?

joshsmith commented on web/views/stripe_customer_view.ex:
```diff
@@ -3,7 +3,29 @@ defmodule CodeCorps.StripeCustomerView do
+  def id_from_stripe(stripe_customer, %Plug.Conn{assigns: %{current_user: current_user}}) do
+    if stripe_customer.user_id == current_user.id, do: stripe_customer.id_from_stripe, else: ""
+  end
+  def id_from_stripe(_stripe_customer, _conn), do: ""
```
I don't think this is what I meant. I meant hiding the relationship from the user view if they're not authed.
